pkg/controller/vsphere/reconciler_test.go

```diff
@@ -95,130 +96,234 @@ func TestClone(t *testing.T) {
+					t.Fatal(err)
 				}
-				raw, err := vsphereapi.RawExtensionFromProviderSpec(&providerSpec)
+				dss, err := hostSystem.ConfigManager().DatastoreSystem(context.Background())
 				if err != nil {
 					t.Fatal(err)
 				}
-				return &machinev1.Machine{
-					ObjectMeta: metav1.ObjectMeta{
-						UID:       "3",
-						Name:      "customFolder",
-						Namespace: namespace,
-					},
-					Spec: machinev1.MachineSpec{
-						ProviderSpec: machinev1.ProviderSpec{
-							Value: raw,
-						},
-					},
+				dir, err := ioutil.TempDir("", fmt.Sprintf("tmpdir"))
+				if err != nil {
+					t.Fatal(err)
+				}
+				_, err = dss.CreateLocalDatastore(context.Background(), "datastore1", dir)
+				if err != nil {
+					t.Fatal(err)
+				}
+				_, err = dss.CreateLocalDatastore(context.Background(), "datastore2", dir)
+				if err != nil {
+					t.Fatal(err)
 				}
 			},
-			expectError: false,
-			cloneVM:     true,
 		},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.testCase, func(t *testing.T) {
-			machineScope, err := newMachineScope(machineScopeParams{
-				client:  fake.NewFakeClientWithScheme(scheme.Scheme, &credentialsSecret),
+			if tc.setupFailureCondition != nil {
+				tc.setupFailureCondition()
+			}
+
+			machineScope := machineScope{
 				Context: context.TODO(),
-				machine: tc.machine(t),
-			})
-			if err != nil {
-				t.Fatal(err)
+				machine: &machinev1.Machine{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "test",
+						Namespace: "test",
+						Labels: map[string]string{
+							machinev1.MachineClusterIDLabel: "CLUSTERID",
+						},
+					},
+				},
+				providerSpec:   &tc.providerSpec,
+				session:        session,
+				providerStatus: &vsphereapi.VSphereMachineProviderStatus{},
+				client:         fake.NewFakeClientWithScheme(scheme.Scheme, &credentialsSecret),
 			}
+
 			machineScope.providerSpec.Template = vm.Name
 
-			if _, err := clone(machineScope); (err != nil) != tc.expectError {
-				t.Errorf("Got: %v. Expected: %v", err, tc.expectError)
+			if tc.machineName != "" {
+				machineScope.machine.Name = tc.machineName
 			}
-			if tc.cloneVM {
-				model.Machine++
+
+			_, err := clone(&machineScope)
+
+			if tc.expectedError != nil {
+				if err == nil {
+					t.Fatal("clone() was expected to return error")
+				}
+				if err.Error() != tc.expectedError.Error() {
+					t.Fatalf("expected: %v, got %v", tc.expectedError, err)
+				}
+			} else {
+				if err != nil {
+					t.Fatalf("clone() was not expected to return error: %v", err)
+				}
 			}
-			if model.Machine != model.Count().Machine {
-				t.Errorf("Unexpected number of machines. Expected: %v, got: %v", model.Machine, model.Count().Machine)
+
+			if tc.cloneVM {
+				vms, err := session.Finder.VirtualMachineList(context.Background(), "*")
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				vmsCount++
+				if vmsCount != len(vms) {
+					t.Errorf("Unexpected number of machines. Expected: %v, got: %v", vmsCount, len(vms))
+				}
```

Comment:
This looks a bit odd, should there be a loop or something here?

Reply:
What do you mean by `loop`? That part of test checks that instance was created after calling `clone()` by comparing the number of instances before and after the call.

Reply:
So the loop here is the `for _, tc := range testCases`, I wonder if there is a way to check that the new VM was created without having a count specifically? This `vmsCount` is shared across multiple subtests, so if one of them fails for some reason, that could have a knock on effect for the other subtests, so we would potentially have difficulty tracking down the problem. Ideally each subtest should share as little as possible/nothing if possible, so there isn't dependencies between them.

Does `clone` return the newly created VM? Could we check that exists in the VMList instead?